src/routes/documents/[id]-[slug]/annotate/+page.svelte

```diff
@@ -0,0 +1,192 @@
+
+  // pagination
+  function next() {
+    $currentPage = Math.min($currentPage + 1, document.page_count);
+    scrollToPage($currentPage);
+    replaceState(pageHashUrl($currentPage), {});
+  }
+
+  function previous() {
+    $currentPage = Math.max($currentPage - 1, 1);
+    scrollToPage($currentPage);
+    replaceState(pageHashUrl($currentPage), {});
+  }
+
+  function gotoPage(n: number) {
+    $currentPage = n;
+    scrollToPage($currentPage);
+    replaceState(pageHashUrl($currentPage), {});
+  }
+
+  function onHashChange(e: HashChangeEvent) {
+    const { hash } = new URL(e.newURL);
+    $currentPage = pageFromHash(hash);
+    scrollToPage($currentPage);
+
+    const noteId = noteFromHash(hash);
+    if (noteId) {
+      $activeNote = document.notes.find((note) => note.id === noteId);
+    }
+  }
+
+  /**
+   * Generate a default zoom, based on mode
+   * @param mode
+   */
+  function getDefaultZoom(
+    mode: ViewerMode,
+  ): number | Sizes | "width" | "height" {
+    if (mode === "document") {
+      return "width";
+    }
+
+    return 1;
+  }
+
+  /**
+   * Generate zoom levels based on mode, since each zooms in a slightly different way
+   */
+  function getZoomLevels(mode: ViewerMode): (string | number)[][] {
+    if (mode === "document") {
+      return [
+        ["width", $_("zoom.fitWidth")],
+        ["height", $_("zoom.fitHeight")],
+        [0.5, "50%"],
+        [0.75, "75%"],
+        [1, "100%"],
+        [1.25, "125%"],
+        [1.5, "150%"],
+        [2, "200%"],
+      ];
+    }
+
+    return [];
+  }
+
+  // for typescript
+  function zoomToScale(zoom: any): number | "width" | "height" {
+    if (zoom === "width" || zoom === "height") {
+      return zoom;
+    }
+
+    return +zoom || 1;
+  }
+</script>
+
+<svelte:window on:hashchange={onHashChange} />
+
+<svelte:head>
+  <link
+    rel="preload"
+    href={asset_url.href}
+    as="fetch"
+    crossorigin="anonymous"
+    type="application/pdf"
+  />
+</svelte:head>
+
+<ContentLayout>
+  <PDF {document} {asset_url} scale={zoomToScale(zoom)} />
+
+  <PageToolbar slot="footer">
+    <svelte:fragment slot="left">
+      <div>
+        <!-- additional controls here -->
+      </div>
+    </svelte:fragment>
+    <svelte:fragment slot="center">
+      <Paginator
+        goToNav
+        on:goTo={(e) => gotoPage(e.detail)}
+        on:next={next}
+        on:previous={previous}
+        bind:page={$currentPage}
+        totalPages={document.page_count}
+        has_next={$currentPage < document.page_count}
+        has_previous={$currentPage > 1}
+      />
+    </svelte:fragment>
+
+    <svelte:fragment slot="right">
+      {#if zoomLevels.length}
+        <label class="zoom">
+          {$_("zoom.zoom")}
+          <select name="zoom" bind:value={zoom}>
+            {#each zoomLevels as [value, label]}
+              <option {value}>{label}</option>
+            {/each}
+          </select>
+        </label>
+      {/if}
+    </svelte:fragment>
+  </PageToolbar>
```

Comment:
At risk of being pedantic but for the sake of establishing strong reuse patterns, I want to quickly say why this markup _should not_ be consolidated and shared between the three pages, while the logic above should. Since the markup relates directly to what's rendered on the page, mostly reuses very standard components like `Paginator`, and can link together common functions with common components, duplication here is not desirable since it would lock up our UI in an undesirable, "uncomposed" manner.

That being said—I think that the "ViewerZoomControl" shared between these files could easily be refactored into a component, but not as necessary here since zoom controls could depend on the page.